### PR TITLE
Bug 1590535 - Update background.js and add end-to-end test.

### PR DESCRIPTION
**Pull request by 0c0w3**
This is done and ready for review except for a question about the
engagement event telemetry pref, the one that caused us problems
in the top-sites experiment.  I need to look into whether we need
to use Marco's experiment API from top sites again or can rely on
changes that Shane made in the meantime.  But this is ready for
first review.

### CODE REVIEW

File: src/background.js [hunk 1]
```diff
@@ -4,62 +4,303 @@
...
+ */
+async function unenroll() {
+  await browser.tabs.onActivated.removeListener(onTabActivated);
+  await browser.webNavigation.onCompleted.removeListener(onWebNavigation);
+  await browser.urlbar.onBehaviorRequested.removeListener(onBehaviorRequested);
+  await browser.urlbar.onResultsRequested.removeListener(onResultsRequested);
+  await browser.urlbar.onResultPicked.removeListener(onResultPicked);
+  await browser.webNavigation.onBeforeNavigate.removeListener(onBeforeNavigate);
+  await browser.windows.onFocusChanged.removeListener(onWindowFocusChanged);
+  await browser.experiments.urlbar.engagementTelemetry.clear({});
```
Comment (mak77) [inline src/background.js]: IT may be safer to do this as first thing, and remove listeners later.

**Review by mak77 — APPROVED** [review]
LGTM, nothing major. If I understand correctly we still need telemetry fixes on top.

File: src/background.js (hunk 1 above)
Comment (0c0w3) [reply src/background.js]: Seems reasonable, I'll update.

**Comment by 0c0w3** [comment]
> LGTM, nothing major. If I understand correctly we still need telemetry fixes on top.

If you mean the questions I had in the original PR comment about whether to use `browser.experiments.urlbar.engagementTelemetry`, I fixed that in c17a426. I autolanded the `browser.telemetry` keyed telemetry bug today, and the `browser.experiments.urlbar.engagementTelemetry` is still open but should land soon. Let me know if you meant something else by telemetry fixes.

**Comment by 0c0w3** [comment]
And I still need to copy api.js and schema.js from mozilla-central to this repo, of course.

**Pull request merged**